### PR TITLE
Fix exception in Urllib3 when dealing with filelike body.

### PR DESCRIPTION
**Pull request by isra17**
### Description

Only use `len` on `body` if it's not a filelike object.

Fixex https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1235

### How Has This Been Tested?

tox -e test-instrumentation-urllib3

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py:321]: How is this testing the fix? am I missing something?

Comment (isra17) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py:321]: The assert itself is not really useful, but the test fail without the fix (Raise an error because we try to call len on a BytesIO).

Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py:321]: I didn't notice. Well, in that case, we would have to use something like `getbuffer.nbytes` for the size?

Comment (isra17) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py]: I did a quick refactoring of the tests and added a special case for `io.BytesIO`.

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py]: Consider using the format of the tests like [this](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/7acc336ccac3df54b70452577365f8fcdfc80276/instrumentation/opentelemetry-instrumentation-tornado/tests/test_metrics_instrumentation.py#L108)
I think that it's more readable and clear what the test verifies each line, and easier to edit in the future.

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py:16]: Look like an unnecessary import

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:274]: Great fix!

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py
Comment (isra17) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py]: Done, I moved all the metrics test into their own test file.

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py
Comment (isra17) [reply instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:274]: Thanks!

**Comment by ocelotl** [comment]
Please resolve conflicts

**Comment by srikanthccv** [comment]
Please fix the conflicts

**Comment by isra17** [comment]
Done again.

**Review by shalevr — APPROVED** [review]

**Comment by isra17** [comment]
Seems like the test broke since the first commit. Pushed a fixed version.

**Comment by gtmanfred** [comment]
@shalevr @srikanthccv  any chance this could be merged? it is preventing us from migrating from opentracing to opentelemetry.

**Comment by srikanthccv** [comment]
It can't be merged unless all the requirements are met. The author didn't let maintainers make edits, which prevents me from updating (out-of-date with base) and merging it.

**Comment by isra17** [comment]
> It can't be merged unless all the requirements are met. The author didn't let maintainers make edits, which prevents me from updating (out-of-date with base) and merging it.

I can't give edit access to the PR since the fork is being owned by an org and not a personnal account. Let me rebase again and fix conflict. If there's conflict again in the future I would suggest you just cherry-pick my commit and open your own pull request. I understand maintaining open-source software is thankless work (And I thank you for otel!), but this PR have many times the amount of work it actually required due to at least 5 rebase with conflict that appears each time right before merging.

**Comment by isra17** [comment]
My rebase conflicted with https://github.com/open-telemetry/opentelemetry-python-contrib/commit/a7bd56354bffb6804c47b363875963ca407a8bb9 which did a similar thing than my PR (moved all the test to `test_urllib3_metrcs.py`, but changed the test to use HttpTestBase while my PR worked on this comment https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1399#discussion_r1015784305 and pretty rewrite the whole test suite. I kept my rewrite instead of the `master` version, however I can't use the HTTP server since it only support `GET` request with a specific payload while my tests requires more use-cases.

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_metrics.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_metrics.py:60]: This function moved to test_base.py, You can get it using self.

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_metrics.py
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_metrics.py:51]: Same

**Comment by isra17** [comment]
Ok, finally fixed the test with the new base utils `assert_...`. Should be green now!

**Comment by srikanthccv** [comment]
> > It can't be merged unless all the requirements are met. The author didn't let maintainers make edits, which prevents me from updating (out-of-date with base) and merging it.
> 
> I can't give edit access to the PR since the fork is being owned by an org and not a personnal account. Let me rebase again and fix conflict. If there's conflict again in the future I would suggest you just cherry-pick my commit and open your own pull request. I understand maintaining open-source software is thankless work (And I thank you for otel!), but this PR have many times the amount of work it actually required due to at least 5 rebase with conflict that appears each time right before merging.


I wouldn't mind somebody else creating the PR for the same fix, getting that merged, and then closing this. You may have n number of reasons not to give edit access, even if it wasn't an org account. I mentioned that because the last time I wanted to merge this two weeks ago, this was not up to date with the base branch. Many PRs in this repo let the maintainer make edits (for such cases) and merge quickly. I also enabled the auto-merge, but that didn't help either. I can't hold other PRs ready to merge because this needs rebasing etc...

**Comment by shalevr** [comment]
LGTM, thank you

**Comment by gtmanfred** [comment]
Thank you very much both of you! we really appreciate it.

**Pull request merged**
